src/src/Controller/CustomerSyncController.php

```diff
@@ -0,0 +1,266 @@
+        if (!$unsubscribers_synchronized) {
+            echo 'Failed to synchronize unsubscribers from Smaily!';
+
+            Logger::logMessageWithSeverity('Customer sync failed - unsubscribers are not removed', 1);
+
+            return false;
+        }
+
+        // Don't sync customers if failed to remove unsubscribers.
+        $subscribers_synchronized = $this->sendSubscribersToSmaily(self::SUBSCRIBERS_BATCH_LIMIT);
+        if (!$subscribers_synchronized) {
+            echo 'Failed to synchronize subscribers from the store!';
+
+            Logger::logMessageWithSeverity('Customer sync failed - failed to send subscribers to Smaily', 1);
+
+            return false;
+        }
+
+        echo 'Subscribers synchronization finished!';
+
+        return true;
+    }
+
+    /**
+     * Get unsubscribers from Smaily and change the subscription status to unsubscribed in store.
+     *
+     * @param int $limit limit request size
+     *
+     * @return bool success status
+     */
+    private function removeUnsubscribers($limit = 1000)
+    {
+        $offset = 0;
+
+        while (true) {
+            $response = $this->api->listUnsubscribers($limit, $offset);
+
+            // Stop if error.
+            if ($response->getStatusCode() !== 200) {
+                Logger::logErrorWithFormatting('Failed fetching unsubscribers.');
+
+                return false;
+            }
+
+            $body = json_decode($response->getBody()->getContents(), true);
+
+            // Stop if no more subscribers.
+            if (empty($body)) {
+                break;
+            }
+
+            // Remove subscribed status for unsubscribers.
+            $subscriberCollectionModel = new SubscriberCollection();
+            $result = $subscriberCollectionModel->batchUnsubscribeByEmail(
+                array_map(
+                    function ($item) {
+                        return $item['email'];
+                    },
+                    $body,
+                )
+            );
+
+            // Stop if query fails.
+            if (!$result) {
+                Logger::logErrorWithFormatting('Failed removing subscribed status for unsubscribers.');
+
+                return false;
+            }
+
+            // Smaily API call offset is considered as page number, not SQL offset!
+            ++$offset;
+        }
+
+        return true;
+    }
+
+    /**
+     * Send store subscribers data to Smaily.
+     *
+     * @param int $limit subscriber request batch limit
+     *
+     * @return bool success status
+     */
+    public function sendSubscribersToSmaily(int $limit): bool
+    {
+        $offset = 0;
+        $subscriberCollection = new SubscriberCollection();
+
+        $additionalFields = json_decode($this->configuration->get('SMAILY_SYNCRONIZE_ADDITIONAL'), true);
+        $selectedFields = array_keys(array_filter($additionalFields));
+
+        while (true) {
```

Comment:
Looks like generator.


```
function getSubscribers($limit) {
    
    $subscribers = [];
    $offset = 0;
    do {
        $subscribers = fetchSubscribers($limit, $offset); //implement me
        $offset++;
        if ($subscribers === false) {
             Logger::logErrorWithFormatting('Failed retrieving newsletter subscribers from DB.');
             throw new Exception('query fails');
            
        }
        foreach($subscribers as $subsriber) {
            yield $subsriber;
        }
    } while($subscribers)
}
```



```
function sendSubscribersToSmaily($limit) {
    try {
        foreach(getSubscribers($limit) as $subscriber) {
        $payload[] = $this->generateSubscriberPayload($subscriber, $selectedFields);
    } catch(Exception $e) {
        return false;
    }
}

```

Reply:
I would keep the same pattern as is as I want batching. Probably could get batching to work with the generator pattern also, but this might increase the complexity required here. The `do-while` loop might read better, but it might also be a matter of preference.